File: src/main/java/io/bloco/faker/FakerComponent.java
package io.bloco.faker;

import java.lang.reflect.InvocationTargetException;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;

import io.bloco.faker.helpers.RandomHelper;
import io.bloco.faker.helpers.StringHelper;

public abstract class FakerComponent {

    private static final String DIGIT_SYMBOL = "#";
    private static final String PARSE_REGEXP = "\\#\\{(.+?)\\}";

    private final FakerData data;
    protected final RandomHelper randomHelper;
    protected final StringHelper stringHelper;

    public FakerComponent(FakerData data) {
        this.data = data;
        this.randomHelper = new RandomHelper();
        this.stringHelper = new StringHelper();
    }

    public String getKey() {
        return stringHelper.camelToSnake(this.getClass().getSimpleName());
    }

    protected String fetch(String key) {
        String[] keys = key.split("\\.");

        List list;
        if (keys.length == 2) {
            list = getList(keys[0], keys[1]);
        } else {
            list = (List) getMap(keys[0], keys[1]).get(keys[2]);
        }

        return sampleFromList(list);
    }

    protected String numerify(String input) {
        return stringHelper.replaceMethod(input, DIGIT_SYMBOL, new StringHelper.StringReplacer() {
            @Override
            public String replaceWith(Matcher matcher) {
                return randomHelper.digit();
            }
        });
    }

    protected String parse(String input) {
        return stringHelper.replaceMethod(input, PARSE_REGEXP, new StringHelper.StringReplacer() {
            @Override
            public String replaceWith(Matcher matcher) {
                String key = matcher.group(1);
                return call(key);
            }
        });
    }

    protected String call(String key) {
        if (key.contains(".")) {
            String[] keys = key.split("\\.");
            return data.getComponentByKey(keys[0]).callMethod(keys[1]);
        } else {
            return callMethod(key);
        }
    }

    protected String getSeparator() {
        return (String) data.get("separator");
    }

    protected String sampleFromList(List options) {
        Object option = randomHelper.sample(options);

        if (option instanceof String) {
            return (String) option;
        } else if (option instanceof List) { // List of lists
            return (String) randomHelper.sample((List) option);
        } else {
            throw new UnsupportedOperationException("Unsupported data type");
        }
    }

    protected List getList(String componentKey, String listKey) {
        List list = (List) getComponentData(componentKey).get(listKey);
        if (list == null) {
            throw new UnsupportedOperationException("Unsupported method '" + listKey + "'");
        }
        return list;
    }

    protected <K extends FakerComponent> K getComponent(Class<K> klass) {
        try {
            return klass.getConstructor(FakerData.class).newInstance(data);
        } catch (InstantiationException|NoSuchMethodException|
                IllegalAccessException|InvocationTargetException e) {
            throw new IllegalArgumentException("Unsupported component '" + klass + "'", e);
        }
    }

    private String callMethod(String methodKey) {
        String methodKeyCamel = stringHelper.snakeToCamel(methodKey);
        String value;
        try {
            value = (String) getClass().getDeclaredMethod(methodKeyCamel).invoke(this);
        } catch (NoSuchMethodException|IllegalAccessException|InvocationTargetException e) {
            throw new IllegalArgumentException(
                    "Unsupported method '" + methodKey + "' " +
                            "for component '" + this.getKey() + "'", e);
        }
        return value;
    }

    protected Map<String, Object> getMap(String componentKey, String listKey) {
        Map<String, Object> map = (Map<String, Object>) getComponentData(componentKey).get(listKey);
        if (map == null) {
            throw new UnsupportedOperationException("Unsupported method '" + listKey + "'");
        }
        return map;
    }

    private Map<String, Object> getComponentData(String componentKey) {
        return data.getComponentData(componentKey);
    }
}

File: lib/src/main/kotlin/io/bloco/faker/FakerComponent.kt
package io.bloco.faker

import io.bloco.faker.helpers.RandomHelper
import io.bloco.faker.helpers.StringHelper
import java.lang.reflect.InvocationTargetException
import kotlin.reflect.KClass

abstract class FakerComponent(private val data: FakerData) {
    protected val randomHelper: RandomHelper = RandomHelper()
    protected val stringHelper: StringHelper = StringHelper()

    val key: String
        get() = stringHelper.camelToSnake(this::class.simpleName!!)

    fun fetch(key: String): String {
        val keys = key.split(".")
        val list = if (keys.size == 2) {
            getList(keys[0], keys[1])
        } else {
            getMap(keys[0], keys[1])[keys[2]] as List<*>
        }
        return sampleFromList(list)
    }

    fun numerify(input: String): String {
        return input.replace(DIGIT_SYMBOL.toRegex()) {
            randomHelper.digit()
        }
    }

    fun parse(input: String): String {
        return input.replace(PARSE_REGEXP.toRegex()) { matcher ->
            val key = matcher.groupValues[1]
            call(key)
        }
    }

    fun call(key: String): String {
        return if (key.contains(".")) {
            val keys = key.split(".")
            data.getComponentByKey<FakerComponent>(keys[0]).callMethod(keys[1])
        } else {
            callMethod(key)
        }
    }

    protected val separator: String
        get() = data["separator"] as String

    protected fun sampleFromList(options: List<*>): String {
        return when (val option = randomHelper.sample(options)) {
            is String -> option
            is List<*> -> randomHelper.sample(option) as String // List of lists
            else -> throw UnsupportedOperationException("Unsupported data type")
        }
    }

    protected fun getList(componentKey: String, listKey: String): List<*> {
        return getComponentData(componentKey)[listKey] as List<*>?
            ?: throw UnsupportedOperationException("Unsupported method '$listKey'")
    }

    protected fun <K : FakerComponent> getComponent(klass: KClass<K>): K {
        try {
            return klass.java.getConstructor(FakerData::class.java).newInstance(data)
        } catch (e: InstantiationException) {
            throw IllegalArgumentException("Unsupported component '$klass'", e)
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException("Unsupported component '$klass'", e)
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException("Unsupported component '$klass'", e)
        } catch (e: InvocationTargetException) {
            throw IllegalArgumentException("Unsupported component '$klass'", e)
        }
    }

    private fun callMethod(methodKey: String): String {
        val methodKeyCamel: String = stringHelper.snakeToCamel(methodKey)
        return try {
            javaClass.getDeclaredMethod(methodKeyCamel).invoke(this) as String
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException("Unsupported method '$methodKey' for component '$key'", e)
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException("Unsupported method '$methodKey' for component '$key'", e)
        } catch (e: InvocationTargetException) {
            throw IllegalArgumentException("Unsupported method '$methodKey' for component '$key'", e)
        }
    }

    protected fun getMap(componentKey: String, listKey: String): Map<String, Any> {
        @Suppress("UNCHECKED_CAST")
        return getComponentData(componentKey)[listKey] as Map<String, Any>?
            ?: throw UnsupportedOperationException("Unsupported method '$listKey'")
    }

    private fun getComponentData(componentKey: String): Map<String, Any> {
        return data.getComponentData(componentKey)
    }

    companion object {
        private const val DIGIT_SYMBOL = "#"
        private const val PARSE_REGEXP = "\\#\\{(.+?)\\}"
    }
}
